pyhealth/datasets/base_dataset.py

```diff
@@ -593,3 +603,170 @@ def task_stat(self) -> str:
     def info():
         """Prints the output format."""
         print(INFO_MSG)
+
+class SampleDataset(ABC, Dataset):
+    """Abstract sample dataset class.
+
+    This dataset takes the processed data samples as an input list
+    
+    Args:
+        samples: the processed data samples.
+            E.g., 
+                samples[0] = {
+                    visit_id: 1,
+                    patient_id: 1,
+                    "codnition": ["A", "B"],
+                    "procedure": ["C", "D"],
+                    "label": 1
+                }
+
+    Attributes:
+        samples: Optional[List[Dict]], a list of samples, each sample is a dict with
+            patient_id, visit_id, and other task-specific attributes as key.
+            Default is None.
+        patient_to_index: Optional[Dict[str, List[int]]], a dict mapping patient_id to
+            a list of sample indices. Default is None.
+        visit_to_index: Optional[Dict[str, List[int]]], a dict mapping visit_id to a
+            list of sample indices. Default is None.
+    """
+
+    def __init__(self, samples, dataset_name="dataset"):
+        self.dataset_name: str = dataset_name
+        self.samples: Optional[List[Dict]] = samples
+        self.patient_to_index: Optional[Dict[str, List[int]]] = None
+        self.visit_to_index: Optional[Dict[str, List[int]]] = None
```

Comment:
I think you should call _index_patient() and _index_visit() in the init. You can also add some sanity check set_task() in the BaseDataset.